bionic/flow.py

```diff
@@ -114,22 +114,21 @@ def add_case(self, name, case_key, value):
     def clear_providers(self, names):
         state = self
 
-        # Remember the original protocol for each provider.
-        protocols_by_entity_name = {}
+        # Remember the original provider attributes per entity.
+        attrs_by_entity_name = {}
         for name in names:
             if not state.has_provider(name):
                 continue
             provider = self.get_provider(name)
-            for res_name, res_protocol in zip(
-                    provider.attrs.names, provider.attrs.protocols):
-                protocols_by_entity_name[res_name] = res_protocol
+            attrs_by_entity_name[name] = provider.attrs
 
         # Delete the providers (or fail if not possible).
         state = state.delete_providers(names)
 
         # Recreate an empty version of each provider.
-        for res_name, res_protocol in protocols_by_entity_name.items():
-            state = state.create_provider(res_name, res_protocol)
+        for name, attrs in attrs_by_entity_name.items():
+            for res_name, res_protocol in zip(attrs.names, attrs.protocols):
```

Comment:
If a provider has multiple names, will we end up processing it twice?  Will that cause `create_provider` to throw an exception the second time?